powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/converter/RecoveryPukConverter.java

```diff
@@ -75,77 +51,8 @@ public RecoveryPukConverter(PowerAuthServiceConfiguration powerAuthServiceConfig
-
-                    // Base64-decode PUK hash
-                    final byte[] pukHashBytes = Base64.getDecoder().decode(pukHashFromDB);
-
-                    // Check that the length of the byte array is sufficient to avoid AIOOBE on the next calls
-                    if (pukHashBytes.length < 16) {
-                        logger.error("Invalid encrypted PUK hash format - the byte array is too short");
-                        // Rollback is not required, error occurs before writing to database
-                        throw localizationProvider.buildExceptionForCode(ServiceError.INVALID_KEY_FORMAT);
-                    }
-
-                    // IV is present in first 16 bytes
-                    final byte[] iv = Arrays.copyOfRange(pukHashBytes, 0, 16);
-
-                    // Encrypted PUK hash is present after IV
-                    final byte[] encryptedPukHash = Arrays.copyOfRange(pukHashBytes, 16, pukHashBytes.length);
-
-                    // Decrypt serverPrivateKey
-                    final byte[] decryptedPukHash = aesEncryptionUtils.decrypt(encryptedPukHash, iv, secretKey);
-
-                    // Return decrypted PUK hash
-                    return new String(decryptedPukHash, StandardCharsets.UTF_8);
-
-                } catch (InvalidKeyException ex) {
-                    logger.error(ex.getMessage(), ex);
-                    // Rollback is not required, cryptography methods are executed before database is used for writing
-                    throw localizationProvider.buildExceptionForCode(ServiceError.INVALID_KEY_FORMAT);
-                } catch (GenericCryptoException ex) {
-                    logger.error(ex.getMessage(), ex);
-                    // Rollback is not required, cryptography methods are executed before database is used for writing
-                    throw localizationProvider.buildExceptionForCode(ServiceError.GENERIC_CRYPTOGRAPHY_ERROR);
-                } catch (CryptoProviderException ex) {
-                    logger.error(ex.getMessage(), ex);
-                    // Rollback is not required, cryptography methods are executed before database is used for writing
-                    throw localizationProvider.buildExceptionForCode(ServiceError.INVALID_CRYPTO_PROVIDER);
-                }
-            }
-            default -> {
-                logger.error("Unknown key encryption mode: {}", encryptionMode.getValue());
-                // Rollback is not required, error occurs before writing to database
-                throw localizationProvider.buildExceptionForCode(ServiceError.UNSUPPORTED_ENCRYPTION_MODE);
-            }
-        }
+    public String fromDBValue(final RecoveryPuk pukHash, final long applicationRid, final String userId, final String recoveryCode, final long pukIndex) throws GenericServiceException {
+        return encryptionService.fromDBValue(pukHash, createSecretKeyDerivations(applicationRid, userId, recoveryCode, pukIndex));
```

Comment:
Why is the method name plural in `createSecretKeyDerivations`?

I suggest to use `deriveSecretKey`.

Reply:
🤔 Since the result is a type of `List`, let's rename that to `createSecretKeyDerivationInputs`, because deviation itself is happening in a different place.